### DISCUSSION TITLE
Select of a list always triggering rebuild

### DISCUSSION THREAD

**Discussion by redarflap** [rrousselGit/riverpod#1655] [Q&A] [answered]
Hi, 
i'm sure this is a common issue, but i haven't found anything.

When i try to select a List, a rebuild is triggered every time the provider changes, even if the list isn't modified.
Even the simplified example below is triggering always:

```
final test = ref.watch(
      projectProvider(projectId).select(
        (project) => ["abc", "def"],
     ),
  );
```

I suppose this happens because of the mutability of lists. Is there any way around this?
The only solution i could find is by making the list const, which isn't really an option in the real usecase.

**Comment by rrousselGit** [comment, marked as answer]
Select works by checking == between the previous and new value


Lists don't have a custom ==, so the comparison is always false with the code you gave.

**Reply by redarflap** [reply]
Oh thanks, i was not aware of that.

I solved it by wrapping the list in a helper class. Maybe there is an easier way, but it seems to work fine.

```
class ComparableList<T> {
  final List<T> _list;
  const ComparableList(this._list);
  get list => _list;
  @override
  bool operator ==(dynamic other) =>
      other.runtimeType == runtimeType &&
      const DeepCollectionEquality()
          .equals((other as ComparableList<T>).list, _list);

  @override
  int get hashCode =>
      Object.hash(runtimeType, const DeepCollectionEquality().hash(_list));
}


 final myList = ref
        .watch(
          projectProvider(projectId).select(
            (project) => ComparableList<Item>(
                project.getList()),
          ),
        )
        .list;
```

**Reply by xleon** [reply]
I had the same issue and adapted your ComparableList a bit so instead of having a `comparableList.list` property, the comparableList is a list itself: 

```dart
import 'package:collection/collection.dart';

class ComparableList<T> extends DelegatingList<T> {
  ComparableList(Iterable<T> source) : super(List.unmodifiable(source));

  @override
  bool operator ==(Object other) =>
      other is ComparableList<T> && const DeepCollectionEquality().equals(this, other);

  @override
  int get hashCode => const DeepCollectionEquality().hash(this);
}

extension ComparableListExtension<T> on Iterable<T> {
  ComparableList<T> toComparableList() => ComparableList(this);
}
```
